### PR TITLE
Balance merge to main

### PR DESCRIPTION
**Pull request by Hsaunders603**

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/Drive.java [hunk 1]
```diff
@@ -299,6 +305,58 @@ public CommandBase driveToNearest() {
 
     }
 
+    private enum robotDriveDirection {
```
Comment (msoucy) [inline src/main/java/frc/robot/subsystems/Drive.java]: Do we still need this?

File: src/main/java/frc/robot/subsystems/Drive.java [hunk 2]
```diff
@@ -299,6 +305,58 @@ public CommandBase driveToNearest() {
...
+
+            double velocityThresholdDegreesPerSec = 1.0;
+
+            double angleVelocityDegreesPerSec = map.gyro().getRotation2d().getCos() * pitchVelocity
+                    + map.gyro().getRotation2d().getSin() * rollVelocity;
+            boolean shouldStop = (pigeonGyro.getPitch() < 0.0 &&
+                    angleVelocityDegreesPerSec > velocityThresholdDegreesPerSec)
+                    || (pigeonGyro.getPitch() > 0.0 && angleVelocityDegreesPerSec < -velocityThresholdDegreesPerSec);
+
+            Logger.getInstance().recordOutput("Pitch Velocity:", map.gyro().getRotation2d().getCos() * pitchVelocity);
```
Comment (msoucy) [inline src/main/java/frc/robot/subsystems/Drive.java]: There's no consistency here - space? No space? Capitalization? Please remove the colons as well.

File: src/main/java/frc/robot/subsystems/Drive.java [hunk 3]
```diff
@@ -299,6 +305,58 @@ public CommandBase driveToNearest() {
...
+            Logger.getInstance().recordOutput("Roll Velocity:", map.gyro().getRotation2d().getSin() * rollVelocity);
+            Logger.getInstance().recordOutput("angleVelocity:", angleVelocityDegreesPerSec);
+            Logger.getInstance().recordOutput("pitch", pigeonGyro.getPitch());
+
+            if (shouldStop) {
+                safeState();
+                Logger.getInstance().recordOutput("Driving:", "I'm stopped!");
+
+            } else {
```
Comment (msoucy) [inline src/main/java/frc/robot/subsystems/Drive.java]: `} else if (pigeonGyro.getPitch() > 3) {`

File: src/main/java/frc/robot/subsystems/Drive.java [hunk 4]
```diff
@@ -299,6 +305,58 @@ public CommandBase driveToNearest() {
...
+                } else if (pigeonGyro.getPitch() < -3) {
+                    move(0.0, -0.25, 0.0);
+
+                    Logger.getInstance().recordOutput("Driving:", "backward");
+                    Logger.getInstance().recordOutput("Balanced yet?", false);
+                }
+
+            }
+
+        }).runsUntil(balancedCheck).onEnd(() -> Logger.getInstance().recordOutput("Driving:", "Command Ended"));
```
Comment (msoucy) [inline src/main/java/frc/robot/subsystems/Drive.java]: Colon in log name

**Review by msoucy — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/maps/OffAxis.java
Comment (bot190) [inline src/main/java/frc/robot/maps/OffAxis.java:1]: Let's update to chopshoplib 2023.4.1 and then we can back-out this change and start getting the 3d rotation data from the `PigeonGyro2` class.

File: src/main/java/frc/robot/subsystems/Drive.java [hunk 5]
```diff
@@ -299,6 +305,59 @@ public CommandBase driveToNearest() {
 
     }
 
+    public CommandBase driveUntilTipped() {
+        return cmd().onExecute(() -> {
+            move(0.0, 0.25, 0.0);
+
+        }).runsUntil(() -> Math.abs(pigeonGyro.getPitch()) > 9);
```
Comment (bot190) [inline src/main/java/frc/robot/subsystems/Drive.java]: The target pitch can probably be lower here. And we should probably be combining pitch and roll based on the yaw to ensure we get an accurate value.

File: src/main/java/frc/robot/subsystems/Drive.java [hunk 6]
```diff
@@ -299,6 +305,59 @@ public CommandBase driveToNearest() {
...
+            double angleVelocityDegreesPerSec = map.gyro().getRotation2d().getCos() * pitchVelocity
+                    + map.gyro().getRotation2d().getSin() * rollVelocity;
+            boolean shouldStop = (pigeonGyro.getPitch() < 0.0 &&
+                    angleVelocityDegreesPerSec > velocityThresholdDegreesPerSec)
+                    || (pigeonGyro.getPitch() > 0.0 && angleVelocityDegreesPerSec < -velocityThresholdDegreesPerSec);
+
+            Logger.getInstance().recordOutput("Pitch Velocity", map.gyro().getRotation2d().getCos() * pitchVelocity);
+            Logger.getInstance().recordOutput("Roll Velocity", map.gyro().getRotation2d().getSin() * rollVelocity);
+            Logger.getInstance().recordOutput("Angle Velocity", angleVelocityDegreesPerSec);
+            Logger.getInstance().recordOutput("Pitch", pigeonGyro.getPitch());
```
Comment (bot190) [inline src/main/java/frc/robot/subsystems/Drive.java]: These can be removed. We'll start logging all 3 values in the Drive data layer with the other changes.

File: src/main/java/frc/robot/subsystems/Drive.java [hunk 7]
```diff
@@ -299,6 +305,59 @@ public CommandBase driveToNearest() {
...
+            if (shouldStop) {
+                safeState();
+                Logger.getInstance().recordOutput("Driving", "I'm stopped!");
+
+            } else {
+                if (pigeonGyro.getPitch() > 3) {
+                    move(0.0, 0.25, 0.0);
+
+                    Logger.getInstance().recordOutput("Driving", "forward");
+                    Logger.getInstance().recordOutput("Balanced", false);
```
Comment (bot190) [inline src/main/java/frc/robot/subsystems/Drive.java]: Let's keep `Driving`, though maybe rename it to `AutoBalanceState`? We should get rid of `Balanced` though.

File: src/main/java/frc/robot/subsystems/Drive.java [hunk 8]
```diff
@@ -299,6 +305,59 @@ public CommandBase driveToNearest() {
...
+            Logger.getInstance().recordOutput("Angle Velocity", angleVelocityDegreesPerSec);
+            Logger.getInstance().recordOutput("Pitch", pigeonGyro.getPitch());
+
+            if (shouldStop) {
+                safeState();
+                Logger.getInstance().recordOutput("Driving", "I'm stopped!");
+
+            } else {
+                if (pigeonGyro.getPitch() > 3) {
+                    move(0.0, 0.25, 0.0);
```
Comment (bot190) [inline src/main/java/frc/robot/subsystems/Drive.java]: Let's pull this speed out into a constant (`private final double`).

File: src/main/java/frc/robot/subsystems/Drive.java [hunk 9]
```diff
@@ -299,6 +305,59 @@ public CommandBase driveToNearest() {
...
+    public CommandBase driveUntilTipped() {
+        return cmd().onExecute(() -> {
+            move(0.0, 0.25, 0.0);
+
+        }).runsUntil(() -> Math.abs(pigeonGyro.getPitch()) > 9);
+    }
+
+    public CommandBase balance() {
+
+        PersistenceCheck balancedCheck = new PersistenceCheck(15, () -> Math.abs(pigeonGyro.getPitch()) < 7);
```
Comment (bot190) [inline src/main/java/frc/robot/subsystems/Drive.java]: We should tweak this value, 7 is probably too high. And it should be using the combined pitch/roll value based on the robot's yaw.

File: src/main/java/frc/robot/Robot.java [hunk 10]
```diff
@@ -123,7 +123,7 @@ public void configureButtonBindings() {
         driveController.rightBumper().onTrue(drive.setSpeedCoef(0.25, 0.35)).onFalse(drive.setSpeedCoef(1, 1));
         driveController.leftBumper().onTrue(drive.setSpeedCoef(0.4000000001, 0.5)).onFalse(drive.setSpeedCoef(1, 1));
         driveController.back().onTrue(drive.resetGyroCommand());
-
+        driveController.a().whileTrue(drive.balance());
```
Comment (bot190) [inline src/main/java/frc/robot/Robot.java]: We should probably remove this (or at least map it to a different button). We don't want Angel accidentally activating this.

**Review by bot190 — CHANGES_REQUESTED** [review]
Awesome work on this, I know it was super frustrating, but hopefully it was rewarding seeing it work!

File: src/main/java/frc/robot/Robot.java (hunk 10 above)
Comment (Hsaunders603) [reply src/main/java/frc/robot/Robot.java]: Fixed, it is now right trigger.

File: src/main/java/frc/robot/maps/subsystems/SwerveDriveMap.java [hunk 11]
```diff
@@ -5,7 +5,9 @@
 
 import com.chopshop166.chopshoplib.drive.MockSwerveModule;
 import com.chopshop166.chopshoplib.drive.SwerveModule;
+
 import com.chopshop166.chopshoplib.sensors.gyro.MockGyro;
+import com.chopshop166.chopshoplib.sensors.gyro.PigeonGyro2;
```
Comment (bot190) [inline src/main/java/frc/robot/maps/subsystems/SwerveDriveMap.java]: This should be removed

File: src/main/java/frc/robot/maps/OffAxis.java [hunk 12]
```diff
@@ -32,7 +33,7 @@ public SwerveDriveMap getDriveMap() {
         // Value taken from CAD as offset from center of module base pulley to center
         // of the robot
         final double MODULE_OFFSET_XY = 0.381;
-        final PigeonGyro pigeonGyro = new PigeonGyro(new PigeonIMU(5));
```
Comment (bot190) [inline src/main/java/frc/robot/maps/OffAxis.java]: This should be changed back to the original. OffAxis has a PigeonGyro not a PigeonGyro2.

File: src/main/java/frc/robot/subsystems/Drive.java [hunk 13]
```diff
@@ -37,6 +42,8 @@ public class Drive extends SmartSubsystemBase {
     double speedCoef = 1;
     double rotationCoef = 1;
 
+    private final double balanceSpeed = 0.25;
```
Comment (bot190) [inline src/main/java/frc/robot/subsystems/Drive.java]: This should be BALANCE_SPEED

File: src/main/java/frc/robot/subsystems/Drive.java [hunk 14]
```diff
@@ -299,6 +306,58 @@ public CommandBase driveToNearest() {
...
+    public CommandBase driveUntilTipped() {
+        return cmd().onExecute(() -> {
+            move(0.0, balanceSpeed, 0.0);
+
+        }).runsUntil(() -> Math.abs(this.getTilt()) > 9);
+    }
+
+    public double getTilt() {
+        return this.map.gyro().getRotation2d().getCos() * this.map.gyro().getRotation3d().getY()
+                + this.map.gyro().getRotation2d().getSin() * this.map.gyro().getRotation3d().getY();
```
Comment (bot190) [inline src/main/java/frc/robot/subsystems/Drive.java]: This should be getX right?

File: src/main/java/frc/robot/subsystems/Drive.java [hunk 15]
```diff
@@ -299,6 +306,58 @@ public CommandBase driveToNearest() {
...
+        PersistenceCheck balancedCheck = new PersistenceCheck(15,
+                () -> Math.abs(this.map.gyro().getRotationalVelocity().getY()) < 6);
+        return cmd().onExecute(() -> {
+            Rotation3d rotationVelocity = this.map.gyro().getRotationalVelocity();
+
+            double velocityThresholdDegreesPerSec = 1.0;
+
+            double angleVelocityDegreesPerSec = map.gyro().getRotation2d().getCos() * rotationVelocity.getY()
+                    + map.gyro().getRotation2d().getSin() * rotationVelocity.getX();
+            boolean shouldStop = (rotationVelocity.getY() < 0.0 &&
```
Comment (bot190) [inline src/main/java/frc/robot/subsystems/Drive.java]: getTilt instead of getY here right?

**Review by bot190 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/Drive.java [hunk 16]
```diff
@@ -299,6 +306,56 @@ public CommandBase driveToNearest() {
...
+
+    public double getTilt() {
+        return this.map.gyro().getRotation2d().getCos() * this.map.gyro().getRotation3d().getY()
+                + this.map.gyro().getRotation2d().getSin() * this.map.gyro().getRotation3d().getX();
+    }
+
+    public CommandBase balance() {
+
+        PersistenceCheck balancedCheck = new PersistenceCheck(15,
+                () -> Math.abs(this.map.gyro().getRotationalVelocity().getY()) < 6);
```
Comment (bot190) [inline src/main/java/frc/robot/subsystems/Drive.java]: This should be looking at `getTilt()`.

**Review by msoucy — APPROVED** [review]

File: src/main/java/frc/robot/subsystems/Drive.java [hunk 17]
```diff
@@ -299,6 +307,60 @@ public CommandBase driveToNearest() {
...
+    }
+
+    public CommandBase balance() {
+
+        PersistenceCheck balancedCheck = new PersistenceCheck(50,
+                () -> Math.abs(getTilt()) < 6);
+        return cmd().onExecute(() -> {
+            Rotation3d rotationVelocity = this.map.gyro().getRotationalVelocity();
+
+            double velocityThresholdDegreesPerSec = 4.0;
```
Comment (bot190) [inline src/main/java/frc/robot/subsystems/Drive.java]: This should be a constant (`private final double`) defined at the top of the class.

File: src/main/java/frc/robot/subsystems/Drive.java [hunk 18]
```diff
@@ -299,6 +307,60 @@ public CommandBase driveToNearest() {
...
+                    || (getTilt() > 1 && angleVelocityDegreesPerSec < -velocityThresholdDegreesPerSec);
+
+            Logger.getInstance().recordOutput("Angle Velocity", angleVelocityDegreesPerSec);
+            Logger.getInstance().recordOutput("Pitch", getTilt());
+
+            if (shouldStop) {
+                safeState();
+                Logger.getInstance().recordOutput("AutoBalanceState", "stopped");
+
+            } else if (getTilt() > 4) {
```
Comment (bot190) [inline src/main/java/frc/robot/subsystems/Drive.java]: These should also be constants at the top of the class.

**Comment by msoucy** [comment]
Merge conflicts. Also please clean up imports

**Review by msoucy — APPROVED** [review]

**Review by bot190 — APPROVED** [review]

**Pull request merged**